src/doc/man/cargo-metadata.md

```diff
@@ -50,6 +53,8 @@ The output has the following format:
                  Example: "registry+https://github.com/rust-lang/crates.io-index"
                - "git+URL" for git-based dependencies.
                  Example: "git+https://github.com/rust-lang/cargo?rev=5e85ba14aaa20f8133863373404cb0af69eeef2c#5e85ba14aaa20f8133863373404cb0af69eeef2c"
+               Consumers should generally see it as an opaque string. There is
+               no guarantee of the format across different versions of Cargo.
```

Comment:
I'm wondering if we should maybe be a little more specific about the guidance for the source string. The lines just above this explicitly state the format of the string, so having this statement about it being opaque seems a little confusing since it contradicts what came just before it.

I think we can probably safely commit to source id's having an `<kind>+` prefix, but leaving the rest to be a little less defined.

Perhaps something along the lines of:

> The value in the source after the `+` is not explicitly defined, and may change between versions of Cargo and may not directly correlate to other things, such as registry definitions in a config file. The entire source string should generally be treated as an opaque string to differentiate packages (such as those with the same name, but coming from different sources). New source kinds may be added in the future which will have different `+` prefixed identifiers.

(That wording could probably be cleaned up a little.)

WDYT?

Reply:
I went ahead and removed the entire mention of `<kind>+`. I think if we want to commit to that, we need to list all possible kinds and explain that. Not sure if we want to do it now so remove.

Reply:
I think it already lists all kinds (all kinds that can appear in `cargo metadata`). The other kinds can't be exposed here AFAIK.

Reply:
I can put it back then.

I was wondering `local-registry` and `directory` but they won't show up in `cargo metadata` as they appears only in source replacment scenario. However, should we mention `sparse` as they are already available for alternative registries?

Reply:
Oh, yea, I forgot about sparse. That should probably be added.

Reply:
Added.

Also added a paragraph to clarify the [compatibility](https://github.com/rust-lang/cargo/blob/2c4064a19cedde64b2aee4a9b7c28951f8dcbdc5/src/doc/man/cargo-metadata.md#compatibility). Please tell me if it is too much.